### ISSUE TITLE
A mistake on gradient penalty

### ISSUE THREAD

**Issue by doub7e** [Zeleni9/pytorch-wgan#40]
https://github.com/Zeleni9/pytorch-wgan/blob/master/models/wgan_gradient_penalty.py#L324

`grad_penalty = ((gradients.norm(2, dim=1) - 1) ** 2).mean() * self.lambda_term`

The shape of `gradients` is [batch_size, 1, 32, 32], the above `norm(2, dim=1)` is actually not behaving as you may want (and returns a tensor with shape [batch_size, 32, 32]). This is actually enforcing `gradients` to be closer to 1 **element-wise**. So the gradient and the lipschitz are much bigger than 1.

A potential fix is
`grad_penalty = (((gradients.view(gradients.shape[0], -1) ** 2).sum(dim=1).sqrt() - 1) ** 2).mean() * self.lambda_term`

Correct me if I am wrong. Thx.

**Comment by R-N** [comment]
You're right. The gradient penalty is now stable and the generator loss isn't changing so fast.

**Comment by R-N** [comment]
Ok, well, this adjustment actually adds artifacts to the generated images? Not sure why

**Comment by R-N** [comment]
Tried your code, and yeah, it has artifacts too. I wonder what's wrong

**Comment by R-N** [comment]
Took a look at the official WGAN GP code and [it doesn't have batch-norm](https://github.com/igul222/improved_wgan_training/blob/fa66c574a54c4916d27c55441d33753dcc78f6bc/gan_mnist.py#L61C1-L61C1). So I removed it and the artifacts are gone. It doesn't seem to be better than before the adjustment though.

![img_generatori_iter_800 (2)](https://github.com/Zeleni9/pytorch-wgan/assets/1442761/4e479c74-489c-44a5-989c-0a2d5674d809)
